core/src/main/java/com/linecorp/armeria/common/metric/ServiceNaming.java

```diff
@@ -0,0 +1,122 @@
+        requireNonNull(defaultServiceName, "defaultServiceName");
+        return cfg -> defaultServiceName;
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} that returns the simple name of innermost class from the given service.
+     * It is supposed to have a class name without a period. e.g. {@code HelloService}.
+     *
+     * @see Class#getSimpleName()
+     */
+    static ServiceNaming simpleTypeName() {
+        return cfg -> {
+            Unwrappable unwrappable = cfg.service();
+            for (;;) {
+                final Unwrappable delegate = unwrappable.unwrap();
+                if (delegate != unwrappable) {
+                    unwrappable = delegate;
+                    continue;
+                }
+                return delegate.getClass().getSimpleName();
+            }
+        };
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} that returns the full name of innermost class from the given service.
+     * It will be fully qualified name including the package name separated by a period.
+     * e.g. {@code com.foo.HelloService}.
+     *
+     * @see Class#getName()
+     */
+    static ServiceNaming fullyQualifiedName() {
+        return cfg -> {
+            Unwrappable unwrappable = cfg.service();
+            for (;;) {
+                final Unwrappable delegate = unwrappable.unwrap();
+                if (delegate != unwrappable) {
+                    unwrappable = delegate;
+                    continue;
+                }
+                return delegate.getClass().getName();
+            }
+        };
+    }
+
+    /**
+     * Converts the specified {@linkplain RequestOnlyLog#serviceName() serviceName}
+     * into another service name which is used as a meter tag or distributed trace's span name.
+     *
+     * <p>A naming rule can be set by either {@link ServerBuilder#defaultServiceNaming(ServiceNaming)} or
+     * {@link com.linecorp.armeria.server.ServiceBindingBuilder#defaultServiceNaming(ServiceNaming)}.
+     * One of pre-defined naming rules is able to be used as follows.
+     * <h2>Example</h2>
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .service(...)
+     *   .defaultServiceNaming(ServiceNaming.simpleTypeName())
+     *   .build()
+     * }</pre>
+     *
+     * <h2>Example 2</h2>
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .route().path("/")
+     *   .defaultServiceNaming(ServiceNaming.fullyQualifiedName())
+     *   ...
+     *   .build()
+     * }</pre>
+     *
+     * <p>If customizing is needed out of given rules, a lambda expression would be applied as follows.
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .service(...)
+     *   .defaultServiceNaming(cfg -> {
+     *     return cfg.server().defaultHostname() + cfg.route().patternString();
+     *   })
+     *   .build()
+     * }</pre>
+     */
+    String convert(ServiceConfig cfg);
```

Comment:
Don't we need to add `ServiceRequestContext` as an input parameter?
We might have to access a `RpcRequest` for a service name of an RPC service?
https://github.com/line/armeria/blob/2258e520cd04a75c3925328e0db1501116fdf09f/core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java#L1052-L1064

Reply:
That is also my concern. If we get `ServiceRequestContext` as a parameter for `RpcRequest`, a default service name cannot be determined in constructor. Do you think that `serviceNaming` is called every request?

Reply:
I don't think it is an expensive operation. I am fine to call it for every request.

Reply:
This function is supposed to be used for determining the 'default' service name only. Doesn't RPC services update the actual service name via `RequestLogBuilder.name(String, String)` when the correct service/method name is determined?